{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exam 2 Answer Javier Rubio Serrano"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1:  Controls\n",
    "\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx).  (hint: This will help you decide how to load the data into the database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Locus\n",
      "AT1G01040\n",
      "AT1G01060\n",
      "AT1G01140\n",
      "AT1G01220\n",
      "AT2G03720\n",
      "AT2G03800\n",
      "AT2G04240\n",
      "AT2G05210\n",
      "AT3G02130\n",
      "AT3G02140\n",
      "AT3G02230\n",
      "AT3G02260\n",
      "AT3G02310\n",
      "AT3G02680\n",
      "AT3G02850\n",
      "AT3G02870\n",
      "AT3G03260\n",
      "AT4G14790\n",
      "AT4G15210\n",
      "AT4G15560\n",
      "AT4G15570\n",
      "AT4G15802\n",
      "AT4G15880\n",
      "AT4G16420\n",
      "AT4G16480\n",
      "AT5G10480\n",
      "AT5G10510\n",
      "AT5G11110\n",
      "AT5G11260\n",
      "AT5G11510\n",
      "AT5G12200\n",
      "AT5G13290\n",
      "Locus\n",
      "AT1G01040\n",
      "AT1G01060\n",
      "AT1G01140\n",
      "AT1G01220\n",
      "AT2G03720\n",
      "AT2G03800\n",
      "AT2G04240\n",
      "AT2G05210\n",
      "AT3G02130\n",
      "AT3G02140\n",
      "AT3G02230\n",
      "AT3G02260\n",
      "AT3G02310\n",
      "AT3G02680\n",
      "AT3G02850\n",
      "AT3G02870\n",
      "AT3G03260\n",
      "AT4G14790\n",
      "AT4G15210\n",
      "AT4G15560\n",
      "AT4G15570\n",
      "AT4G15802\n",
      "AT4G15880\n",
      "AT4G16420\n",
      "AT4G16480\n",
      "AT5G10480\n",
      "AT5G10510\n",
      "AT5G11110\n",
      "AT5G11260\n",
      "AT5G11510\n",
      "AT5G12200\n",
      "AT5G13290\n",
      "same sequence\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "germplasm = open (\"Germplasm.tsv\" , \"r\")\n",
    "germplasm.seek(0)\n",
    "for AGI_Germplasm in germplasm.readlines():\n",
    "    AGI_Germplasm = AGI_Germplasm.split(\"\\t\")[0]\n",
    "    print(AGI_Germplasm)\n",
    "germplasm.close()\n",
    "\n",
    "locusGene = open (\"LocusGene.tsv\" , \"r\")\n",
    "locusGene.seek(0)\n",
    "for AGI_locusGene in locusGene.readlines():\n",
    "    AGI_locusGene = AGI_locusGene.split(\"\\t\")[0]\n",
    "    print(AGI_locusGene)\n",
    "locusGene.close()\n",
    "\n",
    "if (AGI_Germplasm == AGI_locusGene):\n",
    "    print(\"same sequence\")\n",
    "else:\n",
    "    print(\"they are not equals, different sequence!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2:  Design and create the database.  \n",
    "* It should have two tables - one for each of the two data files.\n",
    "* The two tables should be linked in a 1:1 relationship\n",
    "* you may use either sqlMagic or pymysql to build the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>information_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>dataexam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>mysql</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>performance_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>sys</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('information_schema',),\n",
       " ('dataexam',),\n",
       " ('mysql',),\n",
       " ('performance_schema',),\n",
       " ('sys',)]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql create database dataexam;\n",
    "%sql show databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql use dataexam;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "4 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>locus</td>\n",
       "        <td>varchar(100)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "        <td>varchar(100)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>phenotype</td>\n",
       "        <td>varchar(300)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>pubmed</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('locus', 'varchar(100)', 'NO', 'PRI', None, ''),\n",
       " ('germplasm', 'varchar(100)', 'NO', '', None, ''),\n",
       " ('phenotype', 'varchar(300)', 'NO', '', None, ''),\n",
       " ('pubmed', 'int(11)', 'NO', '', None, '')]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql CREATE TABLE germplasm (locus VARCHAR(100) NOT NULL PRIMARY KEY, germplasm VARCHAR(100) NOT NULL, phenotype VARCHAR(300) NOT NULL, pubmed INTEGER NOT NULL);\n",
    "%sql DESCRIBE germplasm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "3 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>locus</td>\n",
       "        <td>varchar(100)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>gene</td>\n",
       "        <td>varchar(100)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>proteinLength</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('locus', 'varchar(100)', 'NO', 'PRI', None, ''),\n",
       " ('gene', 'varchar(100)', 'NO', '', None, ''),\n",
       " ('proteinLength', 'int(11)', 'NO', '', None, '')]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql CREATE TABLE locusGene (locus VARCHAR(100) NOT NULL PRIMARY KEY, gene VARCHAR(100) NOT NULL, proteinLength INTEGER NOT NULL);\n",
    "%sql DESCRIBE locusGene\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database.  There are a variety of strategies to accomplish this.  I will give all strategies equal credit - do whichever one you are most confident with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "import csv\n",
    "locusGene0 = open(\"LocusGene.tsv\" , \"r\")\n",
    "locusGene1 = csv.DictReader(locusGene0, delimiter = \"\\t\", quotechar = '\"')\n",
    "\n",
    "germplasm0 = open(\"Germplasm.tsv\" , \"r\")\n",
    "germplasm1 = csv.DictReader(germplasm0, delimiter = \"\\t\", quotechar = '\"')\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='dataexam',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "#connection.autocommit = False\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        \n",
    "        for row in locusGene1:\n",
    "            sql = \"\"\" INSERT INTO locusGene (locus, gene, proteinLength)\n",
    "            VALUES('{}','{}','{}')\"\"\".format(row[\"Locus\"], row[\"gene\"], row[\"proteinLength\"])\n",
    "            cursor.execute(sql)\n",
    "            \n",
    "        for row in germplasm1:   \n",
    "            sql = \"\"\" INSERT INTO germplasm (locus, germplasm, phenotype, pubmed)\n",
    "            VALUES('{}','{}','{}','{}')\"\"\".format(row[\"Locus\"],row[\"germplasm\"], row[\"phenotype\"], row[\"pubmed\"])\n",
    "            cursor.execute(sql)\n",
    "            \n",
    "    connection.commit()\n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    \n",
    "locusGene0.close()\n",
    "germplasm0.close()\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file\n",
    "\n",
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "2. Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule! \n",
    "\n",
    "All reports should be written to **the same file**.  You may name the file anything you wish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "report1 = %sql SELECT * FROM germplasm, locusGene WHERE germplasm.locus = locusGene.locus;\n",
    "report1 = str(report1)\n",
    "\n",
    "#2. Create a joined report that only includes the Genes SKOR and MAA3 \n",
    "report2 = %sql SELECT * FROM germplasm, locusGene WHERE (locusGene.gene = 'MAA3' OR locusGene.gene = 'SKOR') AND germplasm.locus = locusGene.locus;\n",
    "report2 = str(report2)\n",
    "\n",
    "#3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "report3 = %sql SELECT COUNT(*) AS \"Number of entries for each Chromosome\" FROM locusGene;\n",
    "\n",
    "#4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "report4 = SELECT AVG (locusGene.proteinLength) AS \"Average protein length for the genes on each Chromosome\" FROM LocusGene;\n",
    "report4 = str(report4)\n",
    "\n",
    "Reports = open(\"Newfile_Reports_Exam.csv\", \"w\")\n",
    "Reports.write(\"Problem 4 REPORT 1:{}\\nProblem 4 REPORT 2:{}\\nProblem 4 REPORT 3:{}\\nProblem 4 REPORT 4:{}\".format(report1, report2, report3, report4))\n",
    "Reports.close()\n",
    "\n",
    "file_reports = open(\"Newfile_Reports_Exam.csv\", \"r\")\n",
    "print(file_reports.read())\n",
    "file_reports.close()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
